### PR TITLE
parallel NewMachines

### PR DESCRIPTION
**Pull request by mischief**
on my spinning rust,
before: `2015-09-13T01:34:30Z kola: --- PASS: Etcd2Discovery on qemu (41.673s)`
after: `2015-09-13T01:33:10Z kola: --- PASS: Etcd2Discovery on qemu (24.825s)`

### CODE REVIEW

**Comment by mischief** [comment]
on aws, the increase is quite nice.
before: `2015-09-13T01:41:03Z kola: --- PASS: Etcd2Discovery on aws (279.127s)`
after: `2015-09-13T01:43:12Z kola: --- PASS: Etcd2Discovery on aws (88.195s)`

File: platform/platform.go [hunk 1]
```diff
@@ -116,3 +117,47 @@ func transfer(in io.Reader, m Machine, localPath string) error {
...
+			}
+			mchan <- m
+			wg.Done()
+		}()
+	}
+
+	wg.Wait()
+	close(mchan)
+	close(errchan)
+
```
Comment (vcaputo) [inline platform/platform.go]: I'd prefer machs be declared here, in the form machs := []Machine{}

File: platform/platform.go [hunk 2]
```diff
@@ -116,3 +117,47 @@ func transfer(in io.Reader, m Machine, localPath string) error {
...
+	for i := 0; i < n; i++ {
+		ud := userdatas[i]
+		wg.Add(1)
+		go func() {
+			m, err := c.NewMachine(ud)
+			if err != nil {
+				errchan <- err
+			}
+			mchan <- m
```
Comment (vcaputo) [inline platform/platform.go]: You're still sending m even if NewMachine() fails, won't that m be nil?

Comment (mischief) [inline platform/platform.go]: :clap:

File: platform/platform.go [hunk 3]
```diff
@@ -116,3 +117,47 @@ func transfer(in io.Reader, m Machine, localPath string) error {
...
+	close(mchan)
+	close(errchan)
+
+	for m := range mchan {
+		machs = append(machs, m)
+	}
+
+	if firsterr, ok := <-errchan; ok {
+		for _, m := range machs {
+			m.Destroy()
```
Comment (vcaputo) [inline platform/platform.go]: If you intended to potentially get nil machines, you should probably be testing if m is nil before calling Destroy() here.

Comment (mischief) [inline platform/platform.go]: no, it should be a non-nil machine Machine *or* a non-nil error.

**Comment by mischief** [comment]
retest this please

**Comment by mischief** [comment]
retest this please

**Comment by vcaputo** [comment]
lgtm

**Comment by pbx0** [comment]
lgtm

**Pull request merged**